src/app/dynamic-forms/service/print/print.service.ts

```diff
@@ -181,44 +197,47 @@ export class PrintService {
   }
 
 
-  async getProperContent() {
+  /**
+   * Recursively iterates through the content and resolves any outstanding promises for images. Note that the types are "any" because
+   * properly typing this function would be too verbose for the current ticket.
+   */
+  async resolveImagePromises(content: any): Promise<any> {
 
-    let dontpush;
-    const properContent = [];
+    return new Promise(async (resolve, reject) => {
 
-    return new Promise(async (resolve, rej) => {
+      if (Array.isArray(content)) {
+        const output = new Array<any>();
 
-      for (let i = 0; i < this.content.length; i++) {
-        let item = this.content[i];
+        content.forEach(async (item: any) => {
 
-        if (item.text) {
-          properContent.push(item);
-        }
-        else if (item.columns) {
-          for (let k = 0; k < item.columns.length; k++) {
-            let col = item.columns[k];
-            if (col.image) {
-              // wait for the promise to resolve before advancing the for loop
-              try {
-                const resolvedImage = await col.image;
-                col.image = resolvedImage;
-                col.fff = resolvedImage;
-                dontpush = !resolvedImage.includes("data:image");
-                !dontpush && properContent.push(item);
-              } catch (e) {
-                console.error("Error while getting base64 file");
-              }
-            } else if (k === 1) {
-              properContent.push(item)
-            }
-          }
-        } else {
-          properContent.push(item);
+          output.push(await this.resolveImagePromises(item));
+        });
+
+        resolve(output);
+      }
+      else if (content?.image && content.image instanceof Promise) {
+        const result = await content.image;
+
+        if (result?.includes("data:image")) {
+          // Clone any other properties, but ensure that the image property contains the new data
+          resolve(Object.assign({}, content, { image: result }));
         }
-        if (i === this.content.length - 1) {
-          resolve(properContent);
+        else {
+          resolve(Object.assign({}, content, { text: "Could not resolve image", image: null }));
         }
       }
+      else if (content && typeof content === "object") {
+        const output = {};
+
+        Object.keys(content).forEach(async (key: string) => {
+          output[key] = await this.resolveImagePromises(content[key]);
+        });
+
+        resolve(output);
+      }
+      else {
+        resolve(content);
+      }
     });
   }
```

Comment:
Just thought I'd drop a note here and see if I can simplify this a bit cause I noticed the manual promises being created and that to me usually signals we can do a bit of a refactor.. Is there a reason we don't use Promise.all?

In vanilla JS, that would simplify the code to:

```javascript
async resolveImagePromises(content: any): Promise<any> {
  if (Array.isArray(content)) {
    // Handling arrays with Promise.all for concurrent resolution
    return Promise.all(content.map(item => this.resolveImagePromises(item)));
  } else if (content?.image && content.image instanceof Promise) {
    const result = await content.image;
    // Using spread operator for concise object cloning and property overriding
    if (result?.includes("data:image")) {
      return { ...content, image: result };
    } else {
      return { ...content, text: "Could not resolve image", image: null };
    }
  } else if (content && typeof content === "object") {
    const output = {};
    // Handling objects with Promise.all for concurrent resolution
    await Promise.all(Object.keys(content).map(async (key: string) => {
      output[key] = await this.resolveImagePromises(content[key]);
    }));
    return output;
  } else {
    return content;
  }
}
```

But even then I'd say the function is still way too complex! A second pass could lead to something like (gonna skip typing):

```javascript
async resolveImagePromises(content: any): Promise<any> {
  if (Array.isArray(content)) {
    return this._resolveArray(content);
  } else if (content?.image && content.image instanceof Promise) {
    return this._resolveImage(content);
  } else if (content && typeof content === "object") {
    return this._resolveObject(content);
  } else {
    return content;
  }
}

private async _resolveArray(content: any[]): Promise<any[]> {
  return Promise.all(content.map(item => this.resolveImagePromises(item)));
}

private async _resolveImage(content: any): Promise<any> {
  const result = await content.image;
  if (result?.includes("data:image")) {
    return { ...content, image: result };
  } else {
    return { ...content, text: "Could not resolve image", image: null };
  }
}

private async _resolveObject(content: any): Promise<any> {
  const output = {};
  await Promise.all(Object.keys(content).map(async (key: string) => {
    output[key] = await this.resolveImagePromises(content[key]);
  }));
  return output;
}

```